sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestCollectorFactory.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.sqlobject;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSortedSet;
+import org.jdbi.v3.DBI;
+import org.jdbi.v3.Handle;
+import org.jdbi.v3.Something;
+import org.jdbi.v3.sqlobject.customizers.RegisterContainerMapper;
+import org.jdbi.v3.sqlobject.customizers.SingleValueResult;
+import org.jdbi.v3.tweak.CollectorFactory;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.SortedSet;
+import java.util.UUID;
+import java.util.stream.Collector;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+public class TestCollectorFactory {
+
+    private DBI dbi;
+    private Handle h;
+
+    @Before
+    public void setUp() throws Exception {
+        this.dbi = new DBI("jdbc:h2:mem:" + UUID.randomUUID());
+        this.h = dbi.open();
+        h.execute("create table something (id int primary key, name varchar)");
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        h.close();
+    }
+
+    @Test
+    public void testExists() throws Exception {
+        h.execute("insert into something (id, name) values (1, 'Coda')");
+        h.registerCollectorFactory(new MaybeCollectorFactory());
+
+        @SuppressWarnings("unchecked")
+        Maybe<String> rs = h.createQuery("select name from something where id = :id")
+                .bind("id", 1)
+                .mapTo(String.class)
+                .collectInto(Maybe.class);
```

Comment:
This actually makes me question whether we want a `collectInto(Class<R>)` method. The `collect(Collector<T,A,R>)` method can make type safety guarantees, because the `T` parameter can be inferred from the query's type parameter.